# 5 really cool features of Postgres 10
<br /><br /><br /><br /><br />
## <div style="text-align: right">Author: Jakub Wilkowski </div>

# About me
<br />
## Currently:
### Python developer @ [10Clouds](https://10clouds.com/)
<img class="tenc-header__logo" src="https://10clouds.com/wp-content/themes/thegem/dist/images/10clouds-logo.svg" alt="10Clouds" style="background: #000001">
<br /><br />
## Previously:
#### * Database developer in major telecom company
#### * Application development specialist in consulting/finance
#### * MSc in telecommunications

# Agenda
<br />

* A little bit about postgres
* Environment setup
* New features:
  1. Identity columns
  2. Native partitioning
  3. Multicolumn statistics
  4. More parallelism
  5. Full text search support JSON & JSONB columns
* Summary

# Why Postgres?

<br>
<div align="center"><iframe align="center" src="https://giphy.com/embed/c5iMjFfrUFpza" width="480" height="266" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

# Postgres
<br />
* open source
* object RDBMS
* big flexiblity
* extensions!
* a lot of fun

# Setup

## Postgres docker images

In [37]:
!docker pull postgres:9.6
!docker run -p 5430:5432 --name jupy-old-postgres -e POSTGRES_PASSWORD=mysecretpassword -d postgres:9.6

9.6: Pulling from library/postgres
Digest: sha256:318757ed6291e6a1ef86312ac453b9b4a67b48495b59ca2dece909cb0c688c53
Status: Image is up to date for postgres:9.6
ea95e66ea557f1ad3eca203be4ca690ead9d2cd974dac76fe190902241f83305


In [38]:
!docker pull postgres:10
!docker run -p 5431:5432 --name jupy-new-postgres -e POSTGRES_PASSWORD=mysecretpassword -d postgres:10

10: Pulling from library/postgres
Digest: sha256:73a1c4e98fb961bb4a5c55ad6428470a3303bd3966abc442fe937814f6bbc002
Status: Image is up to date for postgres:10
36618fb47a3cc5a0646b32526799686808e50d51c251ad0179f02d998ac37c3e


## Connection to database(s)

In [5]:
%reload_ext sql
connection96="postgresql+psycopg2://postgres:mysecretpassword@localhost:5430/postgres"
connection10="postgresql+psycopg2://postgres:mysecretpassword@localhost:5431/postgres"

In [14]:
%%sql $connection96
select current_setting('server_version');

1 rows affected.


[('9.6.5',)]

In [21]:
%%sql $connection10
select current_setting('server_version')

1 rows affected.


[('10.0',)]

#  1. Identity columns

ID|datetime|user_id|amount|category_id
---|---|---|---|---
101|'2017-01-01'|123456|456|2
102|'2017-01-02'|123412|1000|4
...|...|...|...|...

```postgresql
INSERT INTO foo(
    id,
    datetime, user_id, amount, category_id
    )
SELECT
    (SELECT max(id) + 1 FROM foo),
    '2017-11-03', 11111, -230, 3;
```

## Example (pg10)


In [3]:
%%sql $connection10
CREATE TABLE foo (id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY, val1 INTEGER);
INSERT INTO foo(val1) VALUES (1);

Done.
1 rows affected.


[]

### Create table with identity column
<br />
```postgres
CREATE TABLE foo (
    id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY, 
    val1 INTEGER);
    
INSERT INTO foo(val1) VALUES (1);
```

### Sequence restart
<br />
```postgres
ALTER TABLE foo ALTER COLUMN id RESTART WITH 1000;
INSERT INTO foo(val1) VALUES (2);
```

In [4]:
%%sql $connection10
ALTER TABLE foo ALTER COLUMN id RESTART WITH 1000;
INSERT INTO foo(val1) VALUES (2);

Done.
1 rows affected.


[]

### Creating copy of the table
<br />
```postgres
CREATE TABLE bar (LIKE foo INCLUDING ALL);
INSERT INTO bar(val1) VALUES (3);
INSERT INTO foo(val1) VALUES (4);
```

In [5]:
%%sql $connection10
CREATE TABLE bar (LIKE foo INCLUDING ALL);
INSERT INTO bar(val1) VALUES (3);
INSERT INTO foo(val1) VALUES (4);

Done.
1 rows affected.
1 rows affected.


[]

### Querying foo & bar
<br />
```postgres
SELECT id, val1, 'foo' AS tbl FROM foo
UNION
SELECT id, val1, 'bar' AS tbl FROM bar
ORDER BY val1;```
<br/>

id  | val1 | tbl
----|------|----
1   |1     |foo
1000|2     |foo
1   |3     |bar
1002|4     |foo

In [12]:
%%sql $connection10 
SELECT id, val1, 'foo' as tbl FROM foo
UNION
SELECT id, val1, 'bar' as tbl FROM bar
ORDER BY val1;

4 rows affected.


[(1, 1, 'foo'), (1000, 2, 'foo'), (1, 3, 'bar'), (1001, 4, 'foo')]

## Meanwhile in pg 9.6...
#### (Above steps were repeated using old syntax)


### Querying foo & bar
<br/>
```postgres
SELECT id, val1, 'foo' as tbl FROM foo
UNION
SELECT id, val1, 'bar' as tbl FROM bar
ORDER BY val1;
```
<br/>

id  | val1 | tbl
----|------|----
1   |1     |foo
1000|2     |foo
1001|3     |bar
1002|4     |foo

In [31]:
%%sql $connection96
DROP TABLE IF EXISTS bar;
DROP TABLE IF EXISTS foo;

CREATE TABLE foo (id SERIAL PRIMARY KEY, val1 INTEGER);
INSERT INTO foo(val1) VALUES (1);
ALTER SEQUENCE foo_id_seq RESTART WITH 1000;
INSERT INTO foo(val1) VALUES (2);
CREATE TABLE bar (LIKE foo INCLUDING ALL);
INSERT INTO bar(val1) VALUES (3);
INSERT INTO foo(val1) VALUES (4);

Done.
Done.
Done.
1 rows affected.
Done.
1 rows affected.
Done.
1 rows affected.
1 rows affected.


[]

In [13]:
%%sql $connection96 
SELECT id, val1, 'foo' as tbl FROM foo
UNION
SELECT id, val1, 'bar' as tbl FROM bar
ORDER BY val1;

4 rows affected.


[(1, 1, 'foo'), (1000, 2, 'foo'), (1001, 3, 'bar'), (1002, 4, 'foo')]

```
                        Table "public.bar"
 Column |  Type   |                    Modifiers                     
--------+---------+-------------------------------------------------
 id     | integer | not null default nextval('foo_id_seq'::regclass)
 val1   | integer | 

Indexes:
    "bar_pkey" PRIMARY KEY, btree (id)
```

### Dropping?
<br/>
```postgres 
DROP TABLE foo;
ERROR:  cannot drop table foo because other objects depend on it
DETAIL:  default for table bar column id depends on sequence foo_id_seq
HINT:  Use DROP ... CASCADE to drop the dependent objects too.```

```postgres
DROP TABLE foo CASCADE;
DROP TABLE```

### Inserting again?
<br/>
```postgres 
INSERT INTO bar(val1) VALUES (5);
ERROR:  null value in column "id" violates not-null constraint
DETAIL:  Failing row contains (null, 5).```

<div align="center"><iframe src="https://giphy.com/embed/8FK0n9SIlod7a" width="480" height="360" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

```
      Table "public.bar"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
 val1   | integer | 
Indexes:
    "bar_pkey" PRIMARY KEY, btree (id)```

# 2. Native partitioning

## What is partitioning for?

 * Create another level of abstraction – we want to query only one (master) table
 * The data themselves should be dispatched to different child tables
<br /><br /><br /><br />
[How Do You Fight Smog with Machine Learning? We Tried, and This Is What Happened](https://10clouds.com/blog/machine-learning-smog-app/)

## How it used to be?

 * Create a master table.
 * Create as many inherited tables with datetime constraints as needed.
 * Create indices, keys, and other constraints on child tables.
 * Create a trigger on the master table that will dispatch rows to proper child tables before insert.

## How is it now?

 * Create a master table, **specify partitioning rule**.
 * Create as many ~~(inherited)~~ child tables with datetime constraints as needed.
 * Create indices, keys, and other constraints on child tables.
 * ~~Create a trigger on the master table that will dispatch rows to proper child tables before insert.~~

## Implementation

### 1. Create master table, specify partitioning rule
<br/>
```postgres
CREATE TABLE measurement(
  id INTEGER GENERATED ALWAYS AS IDENTITY,
  datetime TIMESTAMPTZ,
  site_id INTEGER,
  pollutant_id INTEGER,
  value FLOAT)
PARTITION BY RANGE (datetime);
```

### 2. Create a couple of child tables. Define data range limits they should store
<br/>
```postgres
CREATE TABLE measurement_201708
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-08-01') TO ('2017-09-01');

CREATE TABLE measurement_201709
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-09-01') TO ('2017-10-01');

CREATE TABLE measurement_201710
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-10-01') TO ('2017-11-01');
```

### 3. Add all needed keys and indices, for each child table
<br/>
```postgres
ALTER TABLE measurement_201708 ADD PRIMARY KEY (id);
ALTER TABLE measurement_201708 ADD CONSTRAINT fk_measurement_201708_site 
  FOREIGN KEY (site_id) REFERENCES site(id);
CREATE INDEX idx_measurement_201708_datetime 
  ON measurement_201708(datetime);
```

## Messing with partitions

In [23]:
%%sql $connection10
DROP TABLE IF EXISTS measurement;
DROP TABLE IF EXISTS site;
DROP TABLE IF EXISTS pollutant;

CREATE TABLE site (id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY, name TEXT);
INSERT INTO site(name) values ('Marszalkowska'), ('Niepodleglosci'), ('Podlesna'), ('Wokalna');

CREATE TABLE pollutant (id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY, name TEXT);
INSERT INTO pollutant(name) values ('PM10'), ('PM2.5'), ('CO2');

CREATE TABLE measurement(
id INTEGER GENERATED ALWAYS AS IDENTITY,
datetime TIMESTAMPTZ,
site_id INTEGER,
pollutant_id INTEGER,
value FLOAT)
PARTITION BY RANGE (datetime);

CREATE TABLE measurement_201708
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-08-01') TO ('2017-09-01');
CREATE TABLE measurement_201709
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-09-01') TO ('2017-10-01');
CREATE TABLE measurement_201710
PARTITION OF measurement(datetime)
FOR VALUES FROM ('2017-10-01') TO ('2017-11-01');

ALTER TABLE measurement_201708 ADD PRIMARY KEY (id);
ALTER TABLE measurement_201708 ADD CONSTRAINT fk_measurement_201708_site FOREIGN KEY (site_id) REFERENCES site(id);
CREATE INDEX idx_measurement_201708_datetime ON measurement_201708(datetime);

ALTER TABLE measurement_201709 ADD PRIMARY KEY (id);
ALTER TABLE measurement_201709 ADD CONSTRAINT fk_measurement_201709_site FOREIGN KEY (site_id) REFERENCES site(id);
CREATE INDEX idx_measurement_201709_datetime ON measurement_201709(datetime);

ALTER TABLE measurement_201710 ADD PRIMARY KEY (id);
ALTER TABLE measurement_201710 ADD CONSTRAINT fk_measurement_201710_site FOREIGN KEY (site_id) REFERENCES site(id);
CREATE INDEX idx_measurement_201710_datetime ON measurement_201710(datetime);

Done.
Done.
Done.
Done.
4 rows affected.
Done.
3 rows affected.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

### Insert
<br/>
```postgres 
INSERT INTO measurement(
  datetime, 
  site_id, 
  pollutant_id, 
  value)
SELECT 
  '2017-08-01'::TIMESTAMPTZ + ((random()*90)::int) * INTERVAL '1 day',
  (1 + random()*(SELECT max(id)-1 FROM site))::int,
  (1 + random()*(SELECT max(id)-1 FROM pollutant))::int,
  random()
FROM generate_series(1,1000);
```

In [24]:
%%sql $connection10
INSERT INTO measurement(datetime, site_id, pollutant_id, value)
SELECT '2017-08-01'::TIMESTAMPTZ + ((random()*90)::int) * INTERVAL '1 day',
(1 + random()*(SELECT max(id)-1 FROM site))::int,
(1 + random()*(SELECT max(id)-1 FROM pollutant))::int,
random()
FROM generate_series(1,1000);

1000 rows affected.


[]

### Select
<br/>
```postgres
SELECT * FROM measurement 
WHERE datetime BETWEEN '2017-09-20' AND '2017-09-27';
```

In [25]:
%%sql $connection10
SELECT * FROM measurement WHERE datetime BETWEEN '2017-09-20' AND '2017-09-27' limit 5;

5 rows affected.


[(7, datetime.datetime(2017, 9, 21, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 3, 3, 0.194933926220983),
 (37, datetime.datetime(2017, 9, 23, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 3, 2, 0.314075379632413),
 (43, datetime.datetime(2017, 9, 20, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1, 1, 0.374974506907165),
 (45, datetime.datetime(2017, 9, 21, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 4, 1, 0.973794638644904),
 (46, datetime.datetime(2017, 9, 22, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 2, 3, 0.00471486337482929)]

### Explain
<br/>
```postgres
EXPLAIN SELECT * FROM measurement 
WHERE datetime BETWEEN '2017-09-20' AND '2017-09-27';
```
<br/>

QUERY PLAN|
---|
...|
Bitmap Index Scan on idx_measurement_201709_datetime (cost=0.00..4.22 rows=7 width=0)|


In [29]:
%%sql $connection10
EXPLAIN SELECT * FROM measurement WHERE datetime BETWEEN '2017-09-20' AND '2017-09-27' limit 5;

6 rows affected.


[('Limit  (cost=4.22..11.48 rows=5 width=28)',),
 ('  ->  Append  (cost=4.22..14.39 rows=7 width=28)',),
 ('        ->  Bitmap Heap Scan on measurement_201709  (cost=4.22..14.39 rows=7 width=28)',),
 ("              Recheck Cond: ((datetime >= '2017-09-20 00:00:00+00'::timestamp with time zone) AND (datetime <= '2017-09-27 00:00:00+00'::timestamp with time zone))",),
 ('              ->  Bitmap Index Scan on idx_measurement_201709_datetime  (cost=0.00..4.22 rows=7 width=0)',),
 ("                    Index Cond: ((datetime >= '2017-09-20 00:00:00+00'::timestamp with time zone) AND (datetime <= '2017-09-27 00:00:00+00'::timestamp with time zone))",)]

Tell that you can use inspect db to create a model in django with fake migration, write a function that creates new partitions with keys etc., trigger it once a month

Also there should be a big performance boost on inserts vs old approach

# 3. Multicolumn statistics
## aka correlated statistics

## How does postgres estimate number of returned rows?
<br/>
```postgres
SELECT * 
FROM SOME_TABLE 
WHERE col1 = cond1
  AND col2 = cond2
  AND col3 = cond3;
```

$$rows\,to\,retrieve=total\,number\,of\,rows *p_{predicate\,1}*p_{predicate\,2}*p_{predicate\,3}$$

<div align="center"><iframe src="https://giphy.com/embed/3o85xpYXnjNyfScn28" width="480" height="288" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

In [34]:
%%sql $connection10
DROP TABLE IF EXISTS counting_log;
CREATE TABLE counting_log (id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY, datetime TIMESTAMP WITH TIME ZONE, child_id INTEGER, word TEXT);
INSERT INTO counting_log(datetime, child_id, word) 
SELECT current_timestamp, i%1000, 
CASE WHEN i%4=1 THEN 'eeny' 
     WHEN i%4=2 THEN 'meeny' 
     WHEN i%4=3 THEN 'miny' 
     WHEN i%4=0 THEN 'moe' 
     ELSE 'nope' END 
FROM generate_series(1, 1000000) i;
CREATE INDEX idx_counting_log_child_id on counting_log(child_id);
CREATE INDEX idx_counting_log_datetime on counting_log(datetime);
ANALYZE counting_log;

Done.
Done.
1000000 rows affected.
Done.
Done.


[]

## Example

```postgres
CREATE TABLE counting_log (
  id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY, 
  datetime TIMESTAMP WITH TIME ZONE, 
  child_id INTEGER, 
  word TEXT);
```
<br/>

id|datetime|child_id|word|
--|--------|--------|----
1|'2017-11-08 12:00:00'|123|'eeny'
2|'2017-11-08 15:30:00'|130|'meeny'

```postgres
INSERT INTO counting_log(
  datetime,
  child_id, 
  word)
SELECT 
  current_timestamp, 
  i%1000, 
  CASE WHEN i%4=1 THEN 'eeny' 
    WHEN i%4=2 THEN 'meeny' 
    WHEN i%4=3 THEN 'miny' 
    WHEN i%4=0 THEN 'moe' 
    ELSE 'nope' END 
FROM generate_series(1, 1000000) i;
```

In [37]:
%%sql $connection10 
EXPLAIN SELECT datetime FROM counting_log WHERE child_id=123;

4 rows affected.


[('Bitmap Heap Scan on counting_log  (cost=19.95..2697.67 rows=971 width=8)',),
 ('  Recheck Cond: (child_id = 123)',),
 ('  ->  Bitmap Index Scan on idx_counting_log_child_id  (cost=0.00..19.71 rows=971 width=0)',),
 ('        Index Cond: (child_id = 123)',)]

### Estimated 971 rows with child_id=123

In [38]:
%%sql $connection10 
EXPLAIN SELECT datetime FROM counting_log WHERE word='miny';

2 rows affected.


[('Seq Scan on counting_log  (cost=0.00..19643.00 rows=252867 width=8)',),
 ("  Filter: (word = 'miny'::text)",)]

### Estimated 252867 rows with word='miny'

## Let's do some math!

$$rows\,to\,retrieve=total\,number\,of\,rows *p_{predicate\,1}*p_{predicate\,2}\\
=1000000*\frac{971}{1000000}*\frac{252867}{1000000}\approx 245.534$$

In [40]:
%%sql $connection10 
EXPLAIN SELECT datetime FROM counting_log WHERE word='miny' and child_id=123;

5 rows affected.


[('Bitmap Heap Scan on counting_log  (cost=19.77..2699.92 rows=245 width=8)',),
 ('  Recheck Cond: (child_id = 123)',),
 ("  Filter: (word = 'miny'::text)",),
 ('  ->  Bitmap Index Scan on idx_counting_log_child_id  (cost=0.00..19.71 rows=971 width=0)',),
 ('        Index Cond: (child_id = 123)',)]

## The reality

In [49]:
%%sql $connection10 
SELECT count(datetime) FROM counting_log WHERE word='miny' and child_id=123;

1 rows affected.


[(1000,)]

# <div style="text-align: center">245 != 1000

## What if?
<br/><br/>
* We actually wanted to join another table to above results (i.e. children info)?
* ... and because of such big underestimation query planner chose to use nested loop instead of hash join?
* ... and we had more tables to join with **a lot** more data in it?

<div align="center"><iframe src="https://giphy.com/embed/687qS11pXwjCM" width="480" height="480" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

<img src="https://upload.wikimedia.org/wikipedia/en/thumb/8/82/Reddit_logo_and_wordmark.svg/1280px-Reddit_logo_and_wordmark.svg.png" style="background: #FFFFFF">

<img src="http://ww1.prweb.com/prfiles/2017/05/25/14370539/Hacker%20Noon%20-%20how%20hackers%20start%20their%20afternoon%20AMI%20David%20Smooke.jpg">

<img src="https://scontent-frx5-1.xx.fbcdn.net/v/t31.0-8/456226_10150559388837382_1784277255_o.jpg?oh=b089bd5375b2a9ce7ab5c58295595a03&oe=5AAEF011">

<div align="center"><iframe src="https://giphy.com/embed/hFmIU5GQF18Aw" width="343" height="480" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

## Postgres 10 to the rescue!

In [44]:
%%sql $connection10
CREATE STATISTICS st_counting_log_child_id_word ON child_id, word FROM counting_log;

Done.
Done.
5 rows affected.


[('Bitmap Heap Scan on counting_log  (cost=19.92..2690.43 rows=967 width=8)',),
 ('  Recheck Cond: (child_id = 123)',),
 ("  Filter: (word = 'miny'::text)",),
 ('  ->  Bitmap Index Scan on idx_counting_log_child_id  (cost=0.00..19.68 rows=967 width=0)',),
 ('        Index Cond: (child_id = 123)',)]

```postgres
CREATE STATISTICS st_counting_log_child_id_word 
  ON child_id, word FROM counting_log;
```

In [6]:
%%sql $connection10
ANALYZE counting_log;
EXPLAIN SELECT datetime FROM counting_log WHERE word='miny' and child_id=123;

OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 5431?
could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5431?


## Let's look closer at our new statistics

# 4. More parallelism!

## Parallel queries in Postgres so far
<br/><br/>
* Postgres 9.6:
  * Parallel Scans
    * Sequential scan only
  * Parallel Joins
    * Nested loop
    * Hash join
  * Parallel Aggregation
  

## With pg10 we also get:
<br/><br/>
* Postgres 10:
  * Parallel Scans
    * sequential scan only
    * **bitmap heap scan**
    * **index scan**
    * **index-only scan**
  * Parallel Joins
    * nested loop
    * hash join
    * **merge join**
  * Parallel Aggregation 

## New settings

### Minimal size of a table for which parallelism can be triggered
<br/>
```postgres
show min_parallel_table_scan_size;
```
<br/>

min_parallel_table_scan_size| 
--|
8MB|

In [2]:
%%sql $connection10
show min_parallel_table_scan_size

1 rows affected.


[(u'8MB',)]

### Minimal size of a index for which parallelism can be triggered
<br/>
```postgres
show min_parallel_index_scan_size;
```
<br/>

min_parallel_index_scan_size| 
--|
512kB|

In [4]:
%%sql $connection10
show min_parallel_index_scan_size

1 rows affected.


[(u'512kB',)]

### Maximum number of parallel workers to be used
<br/>
```postgres
show max_parallel_workers;
```
<br/>

max_parallel_workers| 
--|
8|

In [5]:
%%sql $connection10
show max_parallel_workers

1 rows affected.


[(u'8',)]

## Example

In [6]:
%%sql $connection10
DROP TABLE IF EXISTS trigonometry;
CREATE TABLE trigonometry AS SELECT i AS arg, sin(i) AS sine, cos(i) AS cosine, tan(i) AS tangent 
FROM generate_series(0, 100000, 0.01) i;

CREATE INDEX idx_trigonometry_arg ON trigonometry(arg);
CREATE INDEX idx_trigonometry_sine ON trigonometry(sine);
CREATE INDEX idx_trigonometry_cosine ON trigonometry(cosine);

Done.
10000001 rows affected.
Done.
Done.
Done.


[]

```postgres
CREATE TABLE trigonometry 
AS 
SELECT 
  i AS arg, 
  sin(i) AS sine, 
  cos(i) AS cosine, 
  tan(i) AS tangent 
FROM generate_series(0, 100000, 0.01) i;

CREATE INDEX idx_trigonometry_arg ON trigonometry(arg);
CREATE INDEX idx_trigonometry_sine ON trigonometry(sine);
CREATE INDEX idx_trigonometry_cosine ON trigonometry(cosine);
```
<br/>

  arg  |         sine         |        cosine        |       tangent        
-------|----------------------|----------------------|----------------------
     0 |                    0 |                    1 |                    0
  0.01 |  0.00999983333416666 |    0.999950000416665 |   0.0100003333466672


### Parallel aggregate (old stuff)

```postgres
EXPLAIN SELECT count(arg) FROM trigonometry WHERE arg > 50000;
```                                         
<br/>

QUERY PLAN|
--------------------------------------------------------|
 Finalize Aggregate  (cost=138908.78..138908.79 rows=1 width=8)|
   ->  Gather  (cost=138908.56..138908.77 rows=2 width=8)|
         Workers Planned: 2|
         ->  Partial Aggregate  (cost=137908.56..137908.57 rows=1 width=8)|
               ->  Parallel Seq Scan on trigonometry  (cost=0.00..134436.34 rows=1388889 width=32)|
                     Filter: (arg > '50000'::numeric)|


In [22]:
%%sql $connection10
EXPLAIN SELECT count(arg) FROM trigonometry WHERE arg > 50000;

10 rows affected.


[(u'Finalize Aggregate  (cost=140633.42..140633.43 rows=1 width=8) (actual time=1129.714..1129.714 rows=1 loops=1)',),
 (u'  ->  Gather  (cost=140633.20..140633.41 rows=2 width=8) (actual time=1129.676..1129.709 rows=3 loops=1)',),
 (u'        Workers Planned: 2',),
 (u'        Workers Launched: 2',),
 (u'        ->  Partial Aggregate  (cost=139633.20..139633.21 rows=1 width=8) (actual time=1125.419..1125.419 rows=1 loops=3)',),
 (u'              ->  Parallel Seq Scan on trigonometry  (cost=0.00..134441.63 rows=2076628 width=8) (actual time=413.910..946.126 rows=1666667 loops=3)',),
 (u"                    Filter: (arg > '50000'::numeric)",),
 (u'                    Rows Removed by Filter: 1666667',),
 (u'Planning time: 0.174 ms',),
 (u'Execution time: 1135.755 ms',)]

### Parallel index scan (new and shiny)

```postgres
EXPLAIN SELECT * FROM trigonometry WHERE arg > 50000;
```
<br/>

QUERY PLAN|
--------------------------------------------------|
 Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..202722.77 rows=4988362 width=32)|
   Index Cond: (arg > '50000'::numeric)|

In [23]:
%%sql $connection10
EXPLAIN SELECT * FROM trigonometry WHERE arg > 50000;

4 rows affected.


[(u'Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..202545.83 rows=4983908 width=32) (actual time=0.079..1444.759 rows=5000000 loops=1)',),
 (u"  Index Cond: (arg > '50000'::numeric)",),
 (u'Planning time: 0.115 ms',),
 (u'Execution time: 1645.573 ms',)]

### Parallel index scan (2nd attempt)

```postgres
SET parallel_setup_cost=100;
EXPLAIN SELECT * FROM trigonometry WHERE arg > 50000;
```
<br/>

QUERY PLAN|
------------------------------|
 Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..202722.77 rows=4988362 width=32)|
   Index Cond: (arg > '50000'::numeric)|


In [24]:
%%sql $connection10
SET parallel_setup_cost=100;
EXPLAIN SELECT * FROM trigonometry WHERE arg > 50000;

Done.
4 rows affected.


[(u'Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..202545.83 rows=4983908 width=32) (actual time=0.070..1223.772 rows=5000000 loops=1)',),
 (u"  Index Cond: (arg > '50000'::numeric)",),
 (u'Planning time: 0.097 ms',),
 (u'Execution time: 1396.401 ms',)]

<div align="center"><iframe src="https://giphy.com/embed/Az1CJ2MEjmsp2" width="480" height="221" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

### Parallel index scan (one more try)

```postgres
SET parallel_setup_cost=1000;
EXPLAIN SELECT arg FROM trigonometry 
        WHERE sine > 0.999 AND arg >100 AND arg < 10000;
```
<br/>

QUERY PLAN|
----------------------------|
 Gather  (cost=1000.43..40550.21 rows=13356 width=8)|
   Workers Planned: 2|
   ->  Parallel Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..38214.61 rows=5565 width=8)|
         Index Cond: ((arg > '100'::numeric) AND (arg < '10000'::numeric))|
         Filter: (sine > '0.999'::double precision)|

In [27]:
%%sql $connection10
SET parallel_setup_cost=1000;
EXPLAIN SELECT arg FROM trigonometry WHERE sine > 0.999 AND arg >100 AND arg < 10000;

Done.
9 rows affected.


[(u'Gather  (cost=1000.43..40533.65 rows=13430 width=8) (actual time=0.516..147.479 rows=14097 loops=1)',),
 (u'  Workers Planned: 2',),
 (u'  Workers Launched: 2',),
 (u'  ->  Parallel Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..38190.65 rows=5596 width=8) (actual time=0.366..142.974 rows=4699 loops=3)',),
 (u"        Index Cond: ((arg > '100'::numeric) AND (arg < '10000'::numeric))",),
 (u"        Filter: (sine > '0.999'::double precision)",),
 (u'        Rows Removed by Filter: 325301',),
 (u'Planning time: 0.286 ms',),
 (u'Execution time: 153.849 ms',)]

## Let's spread some chaos

<div align="center"><iframe src="https://giphy.com/embed/moiWSfviYKNgc" width="480" height="360" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div>

```postgres
SET max_parallel_workers=0;
SET force_parallel_mode=on;
EXPLAIN ANALYZE SELECT arg FROM trigonometry 
                WHERE sine > 0.999 AND arg >100 AND arg < 10000;
```

QUERY PLAN|
------------------|
 Gather  (cost=1000.43..40550.21 rows=13356 width=8) (actual time=1.015..238.902 rows=14097 loops=1)|
   Workers Planned: 2|
   Workers Launched: 0|
   ...|

In [28]:
%%sql $connection10
SET max_parallel_workers =0;
SET force_parallel_mode=on;
EXPLAIN ANALYZE SELECT arg FROM trigonometry WHERE sine > 0.999 AND arg >100 AND arg < 10000;

Done.
Done.
9 rows affected.


[(u'Gather  (cost=1000.43..40533.65 rows=13430 width=8) (actual time=0.269..259.633 rows=14097 loops=1)',),
 (u'  Workers Planned: 2',),
 (u'  Workers Launched: 0',),
 (u'  ->  Parallel Index Scan using idx_trigonometry_arg on trigonometry  (cost=0.43..38190.65 rows=5596 width=8) (actual time=0.073..257.811 rows=14097 loops=1)',),
 (u"        Index Cond: ((arg > '100'::numeric) AND (arg < '10000'::numeric))",),
 (u"        Filter: (sine > '0.999'::double precision)",),
 (u'        Rows Removed by Filter: 975902',),
 (u'Planning time: 0.156 ms',),
 (u'Execution time: 260.346 ms',)]

In [20]:
%%sql $connection10
SET max_parallel_workers = 8;
SET force_parallel_mode=off;

Done.
Done.


[]

# Full text search support JSON & JSONB columns

# Wait, there's even more

* logical replication
* quorum commit for synchronous replication
* XML tables
* SCRAM authentication
* renaming of some functions
* \if \elif \else statements in psql
* ...

# Wrapping it up

# Questions?

# Thank you!